src/server/lsp/outline.ts

```diff
@@ -0,0 +1,456 @@
+        if (outline.children.length == 1) {
+          genOutline(outline.children[0], `${indent} `);
+        } else if (outline.children.length > 1) {
+          for (let i = 0; i < outline.children.length; ++i) {
+            if (i == outline.children.length - 1) {
+              // indent = indent.substr(0, len - 2) + '  ';
+              genOutline(outline.children[i], `${indent}${bottomCorner}`);
+            } else {
+              genOutline(outline.children[i], `${indent}${middleCorner}`);
+            }
+          }
+        }
+    }
+    if (Array.isArray(root.children) && root.children.length > 0)
+      for (const child of root.children) genOutline(child, '');
+    this.outlineStrings[uri] = lines;
+    this.outlinePanelData[uri] = outlineItems;
+    if (this.outlineVersions[uri] === undefined) {
+      this.outlineVersions[uri] = 0;
+    } else {
+      this.outlineVersions[uri] += 1;
+    }
+  };
+
+  highlightCurrentOutlineItem = async () => {
+    if (
+      this.curOutlineItem !== undefined &&
+      this.outlineBuffer !== undefined &&
+      this.curOutlineItem.lineNumber !== undefined &&
+      this.curOutlineItem.startCol !== undefined &&
+      this.curOutlineItem.endCol !== undefined
+    ) {
+      const windows = await workspace.nvim.windows;
+      for (const win of windows) {
+        try {
+          const buf = await win.buffer;
+          if (buf.id === this.outlineBuffer.id) {
+            buf.clearHighlight();
+            win.setCursor([this.curOutlineItem.lineNumber + 1, 0]).catch((e) => {
+              log(e);
+            });
+            buf
+              .addHighlight({
+                hlGroup: 'HighlightedOutlineArea',
+                line: this.curOutlineItem.lineNumber,
+                colStart: this.curOutlineItem.startCol,
+                colEnd: this.curOutlineItem.endCol,
+              })
+              .catch((e) => {
+                log(e);
+              });
+          }
+        } catch (e) {
+          log(e);
+        }
+      }
+    }
+  };
+
+  updateOutlineBuffer = async (uri: string, force = false) => {
+    if (
+      ((this.outlineVersions[uri] === this.outlineVersions_Rendered[uri] && this.outlineVersions[uri] === undefined) ||
+        this.outlineVersions[uri] !== this.outlineVersions_Rendered[uri] ||
+        uri !== this.renderedOutlineUri ||
+        force) &&
+      this.outlineBuffer
+    ) {
+      this.renderedOutlineUri = uri;
+      let content: string[] = [];
+      if (this.outlineStrings[uri]) {
+        this.outlineVersions_Rendered[uri] = this.outlineVersions[uri];
+        content = this.outlineStrings[uri];
+      }
+      await this.outlineBuffer?.length
+        .then(async (len: number) => {
+          if (!this.outlineBuffer) return;
+          if (Number.isInteger(len)) {
+            await this.outlineBuffer.setOption('modifiable', true);
+            if (len > content.length) {
+              await this.outlineBuffer.setLines([], {
```

Comment:
This is what's been causing the flickering I think. We can do a simple two part setLines (instead of wiping all lines) and see if the flickering improves.

Reply:
Ah yeah I don’t think you need to clear it. You should be able to simply replace from 0 to -1 with the content